### ISSUE TITLE
Enable LTO for binary releases

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#8050]
Locally, running with

```rust
[profile.release]
incremental = true
debug = 0 # Set this to 1 or 2 to get more useful backtraces in debugger.
lto = true
codegen-units = 1
```

gives nice 20% speedup (but takes ages to compile). 

I think we should setup CI to run that on release (both for nightly and weekly). I *think* cargo now supports setting profile values via env variables. So we should add `--lto` flag to `xtask dist`, whichs sets the vars before compilation. We should explicetly enable this flag on CI. Locally, `--lto` should not be run. For metrics, we should avoid LTO as well.

Separately, it might be useful to try to find some missing `#[inline]`s.

**Comment by lnicola** [comment]
Also maybe `mimalloc`?

**Comment by matklad** [comment]
Maybe? Last time that didn't work out due to cross-compilation, but, if we are already using containers for CI, we might as well figure that out.

**Comment by matthiaskrgr** [comment]
I think rust-analyzer shares dependencies with other crates like cargo/clippy/rustfmt in the rustc repo and in order for these changes to have effect, all of them would have to be compiled with 1 cgu and monolithic lto..?

**Comment by lnicola** [comment]
This is mostly about release artifacts in this repo, and not the `rustup` binaries.

**Comment by jhpratt** [comment]
Have you considered `lto = "thin"`? It's a decent middle ground that shouldn't take _too_ much longer to compile, while still apparently providing most of the benefits of LTO.

It doesn't look this this is already enabled, so it might be worth giving a shot.

**Comment by mati865** [comment]
Also there are cases when ThinLTO gives better performance boost than Fat LTO.

**Comment by benmkw** [comment]
regarding cross-crate thinLTO (per crate thin lto is on by default afaik) I came across https://bugzilla.mozilla.org/show_bug.cgi?id=1640982 when trying to enable it, the solution to get it to work seems to be `set -x CARGO_PROFILE_RELEASE_LTO thin` which is what I'm using for all builds at the moment.

**Comment by matklad** [comment]
I don't think time to produce release artifacts really matters, so lto=true would be ok. For local builds, we should just stick with what are default cargo settings.

**Comment by matklad** [comment]
locally, `lto = true` changes the build times from 1 minute to 4 minutes, which is quite OK I think. On CI, the difference will be smaller as it is not as parallel to start with.
